File: baekjoon/gold/2655/2655.py
# input
# 5
# 25 3 4
# 4 4 6
# 9 2 3
# 16 2 5
# 1 5 2
# output
# 3
# 5
# 3
# 1

N = int(input())

info = []
info_dict = dict()
for i in range(N):
  a, h, w = map(int, input().split())
  info.append((i, a, h, w))

dp = [[info[i][2], [info[i][0]]] for i in range(N)]
info.sort(key=lambda x: x[3])
for i in range(len(info) - 1):
  idx, a, h, w = info[i]
  for j in range(i + 1, len(info)):
    jdx, b, hh, ww = info[j]
    if info[i][1] > info[j][1] or info[i][3] > info[j][3]:
      continue
    if dp[jdx][0] < dp[idx][0] + hh:
      dp[jdx][0] = dp[idx][0] + hh
      dp[jdx][1] = dp[idx][1] + [jdx]
ret = max(dp, key=lambda x: x[0])
print(len(ret[1]))
for i in ret[1]:
  print(i+1)
  

File: baekjoon/gold/2655/2655.go
// Need to confirm that this problem can be solved by go

package main

import (
	"fmt"
	"sort"
)

func max(a []int) (idx, val int) {
	for i, v := range a {
		if val < v {
			idx, val = i, v
		}
	}
	return idx, val
}

func main() {
	var n int
	var info [][4]int
	var a, h, w int

	fmt.Scan(&n)
	for i := 0; i < n; i++ {
		fmt.Scan(&a, &h, &w)
		info = append(info, [4]int{i, a, h, w})
	}
	dph := make([]int, n)
	dpi := make([][]int, n)
	for i := 0; i < n; i++ {
		dph[i] = info[i][2]
		dpi[i] = []int{info[i][0]}
	}
	sort.Slice(info, func(i, j int) bool {
		return info[i][3] < info[j][3]
	})
	for i := 0; i < n-1; i++ {
		idx := info[i][0]
		for j := i + 1; j < n; j++ {
			jdx, jh := info[j][0], info[j][2]
			if info[i][1] > info[j][1] || info[i][3] > info[j][3] {
				continue
			}
			if dph[jdx] < dph[idx]+jh {
				dph[jdx] = dph[idx] + jh
				dpi[jdx] = append(dpi[idx], jdx)
			}
		}
	}
	idx, _ := max(dph)
	fmt.Println(dph, dpi)
	fmt.Println(len(dpi[idx]))
	for _, v := range dpi[idx] {
		fmt.Println(v + 1)
	}
}
